dev/core#288 : Use the correct membership date for the notification that appear after completing the membership payment in case pre hook is used

omarabuhussein opened the pull request:
Suppose you have an extension that implements **hook_civicrm_pre** on membership entity that add 1 year (or whatever amount of time) to the membership new end date whenever someone completed a pending payment for the membership, Based on this assume the following example : 


1- You created a pending membership starts today (27/7/2018) and ends in 1 year (26/7/2019).
2- You changed this membership payment (contribution) from "pending" to "completed".


Now without the hook implementation as explained above, the following message will appear : 

```
the contribution record has been saved. 
Membership for "CONTACT NAME" has been updated. The membership End Date is July 26th, 2019.
```

and the membership status will be change from Pending to New, And the end date of the membership will be similar to the one shown in the notification message which is 26/7/2019  which is all good.

But when you have the pre hook implemented as suggest above, the membership end date will be 26/7/2020 (since the hook add one year to the end date) but the notification message will still indicate the old end date which is 26/7/2019. This is wrong and the notification message should show the correct end date that reflects the membership actual end date.



## Update 1

After the discussion below, we decided to remove the end date part from the membership notification for the following reasons : 

1- If someone is paying using a price-set for two memberships, then the notification will show the end date of the last membership only.

2- Removing the end date code make the code much simpler and it does not affect the  functionally nor the usability.

eileenmcnaughton commented:
We should get a test on this

eileenmcnaughton commented on CRM/Contribute/BAO/Contribution.php:
```diff
@@ -2008,7 +2008,7 @@ public static function transitionComponents($params, $processContributionObject
             }
 
-            $updateResult['membership_end_date'] = CRM_Utils_Date::customFormat($dates['end_date'],
+            $updateResult['membership_end_date'] = CRM_Utils_Date::customFormat($membership->end_date,
```
The fact we have customFormat happening here is a bit alarming & it suggests to me this isn't the right place & we should be loading the actual end date much closer to where the message is sent from - I see this is being called from 2 places - one being transitionComponentWithReturnMessage( - if this is about what appears in the message then that function would be the right place.

Note this function is on our toxic functions list - there used to be a bot that added a link to the toxic code protocol page whenever a PR touched a toxic function - https://wiki.civicrm.org/confluence/display/CRM/Toxic+Code+Protocol

omarabuhussein replied:
thanks @eileenmcnaughton will take a look

omarabuhussein commented:
> We should get a test on this

How ?!!! Just looking at how big transitionComponents method  is and how many things it load and do makes me cry 😅

eileenmcnaughton commented:
@omarabuhussein reasonable response :-)

In this case is the function you want to test transitionComponentWithReturnMessage ?

eileenmcnaughton commented:
@omarabuhussein actually - if you move the change up into transitionComponentWithReturnMessage I'd see it as a form-level change & be less concerned about a test too.

omarabuhussein commented:
> In this case is the function you want to test transitionComponentWithReturnMessage ?

the problem is transitionComponentWithReturnMessage() calls transitionComponents() directly and it is hard to test transitionComponentWithReturnMessage() in isolation since it is hard to mock transitionComponents() and use the mocked version in transitionComponentWithReturnMessage().. these two methods are so tightly coupled currently.

omarabuhussein commented:
> actually - if you move the change up into transitionComponentWithReturnMessage I'd see it as a form-level change & be less concerned about a test too.

One problem I see in this approach is that I have to load the membership again from db, but let me think about it and I will come back when I have thought of it better.

eileenmcnaughton commented:
@omarabuhussein I see your concern - it's a low traffic function though so it's probably an OK trade off for less reliance on a toxic function. Would need a code comment at minimum for robustness

mattwire commented:
@omarabuhussein What would happen if you replaced https://github.com/civicrm/civicrm-core/blob/master/CRM/Contribute/BAO/Contribution.php#L5194:
`$updateResult = CRM_Contribute_BAO_Contribution::transitionComponents($params);`

with

`civicrm_api3('Contribution', 'completetransaction', $params);`

If that could be proven to work it would be a big step forward towards removing transitionComponents() entirely (it's currently called in 3 places, that would reduce it to two).

It's becoming pretty clear that there are a lot of issues relating to contribution statuses updating other entities and part of the issue is the multiple code-paths.  I've been working through a lot of those issues on https://github.com/civicrm/civicrm-core/pull/11556 and we are making good progress.  There is at least one date-related issue outstanding where some entities fail to renew but no-one has been able to provide me with concrete examples of where it is going wrong.  I know a number of organisations just avoid the core functionality altogether for these reasons but we really need to push to get it consistent at least.

I can see that this PR seems like a trivial fix, but if we keep patching up the old/unreliable code we're never going to get rid of it and it's going to keep taking up dev/review time as we find more issues with it, until someone finally takes the plunge and moves it to Contribution.completetransaction.

eileenmcnaughton commented:
we do have some tests on this form submit I think - note that transition components would still need tp be called for failed & cancelled status in the short term

mattwire commented:
> note that transition components would still need tp be called for failed & cancelled status in the short term

Why? Are we making any changes to the linked entities when the contribution is failed/cancelled?

mattwire commented:
@omarabuhussein Will you have any time to work on this? I think we need to see some code improvement even if it is only small...

omarabuhussein commented:
Thanks both for your notes, I was so busy for a while but I indeed looking to rework on this as soon as possible (hopefully this week).

regarding : 

>What would happen if you replaced https://github.com/civicrm/civicrm-core/blob/master/CRM/Contribute/BAO/Contribution.php#L5194:
$updateResult = CRM_Contribute_BAO_Contribution::transitionComponents($params);
with
civicrm_api3('Contribution', 'completetransaction', $params);
If that could be proven to work it would be a big step forward towards removing transitionComponents() entirely (it's currently called in 3 places, that would reduce it to two).

hmm, this will require going through the path of each one and noting what each path do and seeing if this can really be replaced. this might take time but will do that once  I get back to work on it.

> It's becoming pretty clear that there are a lot of issues relating to contribution statuses updating other entities and part of the issue is the multiple code-paths. I've been working through a lot of those issues on #11556 and we are making good progress. There is at least one date-related issue outstanding where some entities fail to renew but no-one has been able to provide me with concrete examples of where it is going wrong. I know a number of organisations just avoid the core functionality altogether for these reasons but we really need to push to get it consistent at least.
I can see that this PR seems like a trivial fix, but if we keep patching up the old/unreliable code we're never going to get rid of it and it's going to keep taking up dev/review time as we find more issues with it, until someone finally takes the plunge and moves it to Contribution.completetransaction.

Yeah, I would say in general contribution related code is a huge mess, and I really would like to improve things even if I am doing trivial things but most of the time it is really hard, there are different use cases, scenarios and customization built on top all of that and at the same time there is no clear documentation or written specs so it is usually hard to do changes properly.

omarabuhussein commented:
Okay so I found that **membership_end_date** returned from **transitionComponents** method is only used in **transitionComponentWithReturnMessage** method so I removed it and moved the work to **transitionComponentWithReturnMessage**. So currently **transitionComponentWithReturnMessage** will use the contribution id to fetch the membership end date, and while there could be more than one membership but I select the last one as the previous code used to do and because I am not sure what to do instead, but I was thinking of removing the whole date thing from the notification message, what do u think ?.


regarding replacing 

```
$updateResult = CRM_Contribute_BAO_Contribution::transitionComponents($params);
```
with 

```
civicrm_api3('Contribution', 'completetransaction', $params);
```

and yeah as Eileen said, it does not seem to handle failed and cancelled cases. Though I had a headache comparing both and trying to note differences down, but maybe for another day 😅

eileenmcnaughton commented:
"I was thinking of removing the whole date thing from the notification message, what do u think ?."

Yeah - why not. Less complexity & it seems to be no real loss

mattwire commented:
> "I was thinking of removing the whole date thing from the notification message, what do u think ?."

@omarabuhussein Yes! It simplifies old code that we would like to replace anyway - good idea.

omarabuhussein commented:
done

JoeMurray commented:
This seems to mean you are removing the membership dates from renewal messages or something similar. That should be run past partners channel if true, or at least dev channel. User level feedback.

aydun commented:
> This seems to mean you are removing the membership dates from renewal messages or something similar. That should be run past partners channel if true, or at least dev channel. User level feedback.

@omarabuhussein - Can you confirm exactly what is being changed in the latest version of this?  Looks to me that it is just removing the (possibly erroneous) date from the pop-up notification after changing a contribution/membership.  It does not affect other things like renewal emails which is the concern I hear in Joe's comment.

eileenmcnaughton commented:
@aydun yes you are right & reading your comment @JoeMurray's concern might have been a misunderstanding

eileenmcnaughton commented:
Ok - that's actually 3 times I've emailed dev-list & mentioned this & @aydun is the only responder so I think there is not concern about this change. I would also note that @aydun suggests some mis-understanding about the scope of this change was involved. I'm merging this

MegaphoneJon commented:
I have an open PR that I'll modify based on the results of this (membership#9 I believe)

eileenmcnaughton commented:
thanks @MegaphoneJon

omarabuhussein commented:
ops another PR that I missed, sorry for the delay it is too late apparently but as elieen said this only stop showing the dates in the notification and nothing else. thanks a lot anyway
